tests/upgrading/kubernetes_upgrade_test.go

```diff
@@ -0,0 +1,139 @@
+	namegen "github.com/rancher/rancher/tests/framework/pkg/namegenerator"
+	"github.com/rancher/rancher/tests/framework/pkg/session"
+	"github.com/rancher/tfp-automation/config"
+	"github.com/rancher/tfp-automation/framework"
+	cleanup "github.com/rancher/tfp-automation/framework/cleanup"
+	"github.com/rancher/tfp-automation/tests/extensions/provisioning"
+	"github.com/stretchr/testify/require"
+	"github.com/stretchr/testify/suite"
+)
+
+type KubernetesUpgradeTestSuite struct {
+	suite.Suite
+	client             *rancher.Client
+	standardUserClient *rancher.Client
+	session            *session.Session
+	terraformConfig    *config.TerraformConfig
+	clusterConfig      *config.TerratestConfig
+	terraformOptions   *terraform.Options
+}
+
+func (r *KubernetesUpgradeTestSuite) SetupSuite() {
+	testSession := session.NewSession()
+	r.session = testSession
+
+	client, err := rancher.NewClient("", testSession)
+	require.NoError(r.T(), err)
+
+	r.client = client
+
+	enabled := true
+	var testuser = namegen.AppendRandomString("testuser-")
+	var testpassword = password.GenerateUserPassword("testpass-")
+	user := &management.User{
+		Username: testuser,
+		Password: testpassword,
+		Name:     testuser,
+		Enabled:  &enabled,
+	}
+
+	newUser, err := users.CreateUserWithRole(client, user, "user")
+	require.NoError(r.T(), err)
+
+	newUser.Password = user.Password
+
+	standardUserClient, err := client.AsUser(newUser)
+	require.NoError(r.T(), err)
+
+	r.standardUserClient = standardUserClient
+
+	terraformConfig := new(config.TerraformConfig)
+	ranchFrame.LoadConfig("terraform", terraformConfig)
+
+	r.terraformConfig = terraformConfig
+
+	clusterConfig := new(config.TerratestConfig)
+	ranchFrame.LoadConfig("terratest", clusterConfig)
+
+	r.clusterConfig = clusterConfig
+
+	terraformOptions, err := framework.Setup(r.T())
+	require.NoError(r.T(), err)
+
+	r.terraformOptions = terraformOptions
+}
+
+func (r *KubernetesUpgradeTestSuite) TestKubernetesUpgrade() {
+	nodeRolesAll := []config.Nodepool{config.AllRolesNodePool}
+	nodeRolesShared := []config.Nodepool{config.EtcdControlPlaneNodePool, config.WorkerNodePool}
+	nodeRolesDedicated := []config.Nodepool{config.EtcdNodePool, config.ControlPlaneNodePool, config.WorkerNodePool}
+
+	tests := []struct {
+		name      string
+		nodeRoles []config.Nodepool
+		client    *rancher.Client
+	}{
+		{"1 Node all roles " + config.AdminClientName.String(), nodeRolesAll, r.client},
+		{"1 Node all roles " + config.StandardClientName.String(), nodeRolesAll, r.standardUserClient},
+		{"2 nodes - etcd/cp roles per 1 node " + config.AdminClientName.String(), nodeRolesShared, r.client},
+		{"2 nodes - etcd/cp roles per 1 node " + config.StandardClientName.String(), nodeRolesShared, r.standardUserClient},
+		{"3 nodes - 1 role per node " + config.AdminClientName.String(), nodeRolesDedicated, r.client},
+		{"3 nodes - 1 role per node " + config.StandardClientName.String(), nodeRolesDedicated, r.standardUserClient},
+	}
```

Comment:
I see how this could make sense to have statically defined. But maybe we should have an offline discussion about it later. for now it will be fine to leave in, but I feel that this doesn't make the most sense to have in TFP automation.

Reply:
Let's talk about that post this PR. There are additional designs that Josh and I are looking at and that would be better suited for those conversations IMO.